### PR TITLE
WIP: Save path segments

### PR DESCRIPTION
**Pull request by nrc**
DON'T LAND

The idea with this PR is to save a def for every element of a path (we currently know this in name resolution but don't save it). Then we can save it in save-analysis and in an IDE we will be able to jump from the prefix segments of a path to their definition (which is a glaring omission in the RLS right now).

Obviously this is missing the save-analysis parts.

However, I am stuck because the compiler is panicking when bootstrapping std. I'm getting `path resolved multiple times` from [here](https://github.com/rust-lang/rust/blob/master/src/librustc_resolve/lib.rs#L4454). From backtraces this is happening when resolving imports, (`finalize_import` is in the backtrace). However, it looks like that is only called once, from `resolve_crate`, and I think that is the only time I would record the path segments (i.e., I wouldn't when calling `resolve_import` because `record_used` is `false`).

So, questions - am I doing anything stupid here which I shouldn't be doing at all or should be doing differently? Or do you know/can you see any way which we could be resolving the path segments more than once?

cc @petrochenkov @eddyb @jseyfried

### CODE REVIEW

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -286,11 +286,11 @@ impl<'a, 'cl> Resolver<'a, 'cl> {
...
                         if crate_name != keywords::Invalid.name() {
                             // `crate_name` should not be interpreted as relative.
-                            module_path.push(Ident {
+                            module_path.push((Ident {
                                 name: keywords::CrateRoot.name(),
-                                span: source.span,
-                            });
-                            source.name = crate_name;
+                                span: source.0.span,
+                            }, Some(ast::CRATE_NODE_ID)));
```
Comment (nrc) [inline src/librustc_resolve/build_reduced_graph.rs]: This should be a fresh node id, I think, but the id that is getting multiply resolved is not `CRATE_NODE_ID`

**Comment by oli-obk** [comment]
>  I'm getting path resolved multiple times

Do you have more details? Is that happening on the last segment?

**Comment by nrc** [comment]
cc https://github.com/nrc/rls-analysis/issues/109

**Comment by nrc** [comment]
I figured out what was going wrong - when resolving a nested import, we were using the same path prefix, rather than cloning, which was fine before, but now the segments would be resolved multiple times

**Comment by eddyb** [comment]
@nrc Ouch. If we don't hack around it, we could probably use it as a start of a redesign of how imports are resolved, to only resolve the prefix once, and then have the nested imports depend on the prefix import.

**Pull request closed without merging**